### PR TITLE
Various fixes and added debug sync_registers method

### PR DESCRIPTION
**Pull request by wargio**
# DO NOT SQUASH ME

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- Removes the `esil` debugger since it is a big todo
- Removes the unimplemented methods that are defined in `debug.h`
- Adds `const` to `rz_bv_as[_hex]_string`
- Adds `sync_registers` in debug plugin and cleanup brainfuck debugger
- Rewrite of `bochs debugger` and of the `io` plugin
- Removes `rzbochs` subproject

### CODE REVIEW

File: librz/debug/debug.c [hunk 1]
```diff
@@ -1636,7 +1625,7 @@ RZ_API bool rz_debug_is_dead(RzDebug *dbg) {
 	if (!strcmp(dbg->cur->name, "io")) {
 		return false;
 	}
-	bool is_dead = (dbg->pid == -1 && strncmp(dbg->cur->name, "gdb", 3)) || (dbg->reason.type == RZ_DEBUG_REASON_DEAD);
+	bool is_dead = (dbg->pid == -1 && strncmp(dbg->cur->name, "gdb", 3) && strncmp(dbg->cur->name, "bochs", 5)) || (dbg->reason.type == RZ_DEBUG_REASON_DEAD);
```
Comment (XVilka) [inline librz/debug/debug.c]: Could you please extract that `strncmp(gdb) && strncmp(bochs)` comparison into a separate inline function? And also, write a comment about why those in particular.

Comment (XVilka) [inline librz/debug/debug.c]: Could you please extract that `strncmp(gdb) && strncmp(bochs)` comparison into a separate inline function? And also, write a comment about why those in particular.

File: librz/debug/p/debug_bochs.c [hunk 2]
```diff
@@ -1,371 +1,589 @@
...
+ * Bochs is a highly portable open source IA-32 (x86) PC emulator written in C++, that runs on most popular
+ * platforms. It includes emulation of the Intel x86 CPU, common I/O devices, and a custom BIOS.
+ *
+ * This plugin uses the spawned process used by io_bochs plugin to set/unset breakpoints, to get and set the
+ * registers of the CPU.
+ *
+ * The plugin uses the following commands to handle the debugger commands from rizin:
+ *
+ * - `break <addr>` is used to set the breakpoint at a given address.
+ *
```
Comment (XVilka) [inline librz/debug/p/debug_bochs.c:19]: Too many empty lines? Just one is enough, I think

File: librz/debug/p/debug_bochs.c [hunk 3]
```diff
@@ -1,371 +1,589 @@
...
+		r13: 00000000_00000000
+		r14: 00000000_00000000
+		r15: 00000000_00000000
+		rip: 00000000_0000fff0
+		eflags 0x00000002: id vip vif ac vm rf nt IOPL=0 of df if tf sf zf af pf cf
+	*/
+	char *raw = debug_bochs_send_command(bochs, true, "regs\n");
+	char *output = raw ? strstr(raw, "CPU0:") : NULL;
+	if (!output) {
+		eprintf("sync:\n%s\n", raw);
```
Comment (XVilka) [inline librz/debug/p/debug_bochs.c]: Why `eprintf()`? Should it be `RZ_LOG_ERROR()` instead?

File: librz/io/p/io_bochs.c [hunk 4]
```diff
@@ -1,104 +1,264 @@
...
-		bochs_send_cmd(desc, &cmd[1], true);
-		io->cb_printf("%s\n", desc->data);
-	} else if (!strncmp(cmd, "dobreak", 7)) {
-		bochs_cmd_stop(desc);
-		io->cb_printf("%s\n", desc->data);
+static char *io_bochs_command(RzIO *io, RzIODesc *fd, const char *cmd) {
+	RzSubprocess *bochs = fd->data;
+	if (!strncmp(cmd, "pid ", strlen("pid "))) {
+		return NULL;
+	} else {
```
Comment (XVilka) [inline librz/io/p/io_bochs.c]: No need for else here?

File: librz/io/p/io_bochs.c [hunk 5]
```diff
@@ -1,104 +1,264 @@
...
-			" R!:<bochscmd>      - Send a bochs command.\n"
-			" R!dobreak          - pause bochs.\n");
-		lprintf("io_system: Enviando commando bochs\n");
-		bochs_send_cmd(desc, &cmd[1], true);
-		io->cb_printf("%s\n", desc->data);
-	} else if (!strncmp(cmd, "dobreak", 7)) {
-		bochs_cmd_stop(desc);
-		io->cb_printf("%s\n", desc->data);
+static char *io_bochs_command(RzIO *io, RzIODesc *fd, const char *cmd) {
+	RzSubprocess *bochs = fd->data;
```
Comment (XVilka) [inline librz/io/p/io_bochs.c]: Also you can move this line after `strncmp()`

File: librz/io/p/io_bochs.c [hunk 6]
```diff
@@ -1,104 +1,264 @@
...
+		}
+	}
+	size_t written = cur - buf;
+	free(raw);
+	return written == count;
 }
 
-static int __read(RzIO *io, RzIODesc *fd, ut8 *buf, int count) {
+static int io_bochs_read(RzIO *io, RzIODesc *fd, ut8 *buf, int count) {
 	memset(buf, 0xff, count);
```
Comment (XVilka) [inline librz/io/p/io_bochs.c]: Maybe move the `memset()` after the error check?

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/debug/p/debug_bochs.c (hunk 3 above)
Comment (wargio) [reply librz/debug/p/debug_bochs.c]: this one was there for debug. i forgot to remove it.

**Review by XVilka — APPROVED** [review]

**Pull request merged**